### DISCUSSION TITLE
Why the contract address are same

### DISCUSSION THREAD

**Discussion by TheDuyIT** [Cyfrin/foundry-full-course-cu#1628] [Q&A] [answered]
On this lesson, https://updraft.cyfrin.io/courses/solidity/storage-factory/deploying-a-contract-from-a-contract?lesson_format=video (2:53)
After hitting createSimpleStorage button the first time, I figure out the SimpleStorage contract address are the same between mine and the one presenting on the video: 0xd840735F4B6a0d1AF8Fa48EcE560f4778c007397

I am wonder why that's happen?  I thought it must be different address.
Here is my code, same intent but with a little different in naming with the original on video

```javascript
// SPDX-License-Identifier: MIT
pragma solidity 0.8.19;


contract SimpleStorage {
    uint256 myFavoriteNumber = 0;
    // uint256[] favoriteNumbers;
    struct Person {
        uint256 favoriteNUmber;
        string name;
    }
    Person[] public persons;

    mapping (string => uint256) public name2FavoriteNumber;


    function addPerson(string memory _name, uint256 _favoriteNumber) public {
        persons.push(Person(_favoriteNumber, _name));
        name2FavoriteNumber[_name] = _favoriteNumber;
    }

    function store(uint256 _favoriteNumber) public {
        myFavoriteNumber = _favoriteNumber;
    }

    function retrieve() public view returns (uint256) {
        return myFavoriteNumber;
    }
}


contract StorageFactory {
    SimpleStorage public simpleStorage;

    function createSimpleStorageContract() public {
        simpleStorage = new SimpleStorage();    
    }
}
``` 
Thanks you for responding!

**Comment by EngrPips** [comment]
Hello @TheDuyIT, So I went ahead to deploy the exact code you shared on `remix,` and it happens that my contract was deployed to a different address, `0xC3Ba5050Ec45990f76474163c5bA673c244aaECA`. I am unsure why you have the same address as the one in the tutorial. Below is an image that shows my remix deployment 

<img width="1440" alt="Screenshot 2024-04-21 at 08 42 49" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/95760918/f0e23dc0-a364-454f-9907-aaa1cfb36fbc">

**Reply by TheDuyIT** [reply]
I have re-deployed the StorageFactory contract and am unable to reproduce the issue. Unfortunately, I forgot to capture a screenshot when the issue occurred.

**Comment by TheDuyIT** [comment]
I suppose the ENV Remix (Merge) uses some seed addresses (wallet addresses) to simulate the process of generating a contract address. Eventually, my seed address turned out to be the same as the one being used in the video. Since the two codes have the same intent, the payload for creating the contract is also the same (the compiler generates the same bytecodes). Consequently, we end up with the same contract address.

Could someone verify this?

**Reply by EngrPips** [reply]
This makes sense, but I would keep going and not let it slow me down

**Reply by TheDuyIT** [reply, marked as answer]
After continuing to learn following the lesson and conducting some research, I would like to update the answer slightly because the point about **the compiler generating the same bytecodes** is incorrect.

**The contract address is determined based on the person's address creating the contract (sender) and the number of transactions the creator has sent (nonce).** [ref](https://www.quicknode.com/guides/ethereum-development/smart-contracts/an-overview-of-how-smart-contracts-work-on-ethereum) => so **contract bytecode is not a factor**

There is a section called Account for each VM environment. I have kept it default like the one in the tutorial. So with the same wallet address and nonce, the confusion I encountered happens.
Now you can reproduce it by following these steps:
- Deploy a contract in env A
- Switch to another env B (to reset nonce)
- Switch back to environment A and deploy either that contract or any other contract.
The contract address at step 1 and 3 will be the same.

**Reply by TheDuyIT** [reply]
Thank you for joining this discussion! @EngrPips

**Reply by EngrPips** [reply]
Interesting, I mean it is all my pleasure. Keep crushing it, champ
